File: Power Query/400-499/410/PQ_410.py
import pandas as pd

path = "400-499/410/PQ_Challenge_410.xlsx"
input = pd.read_excel(path, usecols="A:D", nrows=20)
test = pd.read_excel(path, usecols="F:I", nrows=20, keep_default_na=False)
test.columns = ["Group", "EliminationOrder", "PersonID", "PersonName"]

rows = []
for group, data in input.groupby("Group", sort=False):
    alive = list(data.PersonID)
    names = dict(zip(data.PersonID, data.PersonName))
    step_size = int(data.StepSize.iloc[0])
    index = 0

    for order in range(1, len(alive) + 1):
        index = (index + step_size - 1) % len(alive)
        person_id = alive.pop(index)
        rows.append([group, order, person_id, names[person_id]])

result = pd.DataFrame(
    rows, columns=["Group", "EliminationOrder", "PersonID", "PersonName"]
)
print(result.equals(test))

File: Power Query/400-499/410/PQ_410.R
library(tidyverse)
library(readxl)

path <- "400-499/410/PQ_Challenge_410.xlsx"
input <- read_excel(path, range = "A1:D21")
test <- read_excel(path, range = "F1:I21")

solve <- function(PersonID, PersonName, StepSize) {
  alive <- PersonID
  names <- setNames(PersonName, PersonID)
  index <- 1

  map_dfr(seq_along(alive), function(order) {
    index <<- ((index + StepSize - 2) %% length(alive)) + 1
    person_id <- alive[index]
    alive <<- alive[-index]
    tibble(
      EliminationOrder = order,
      PersonID = person_id,
      PersonName = unname(names[as.character(person_id)])
    )
  })
}

result <- input %>%
  group_by(Group) %>%
  group_modify(~ solve(.x$PersonID, .x$PersonName, .x$StepSize[[1]])) %>%
  ungroup()

all.equal(result, test)
# TRUE
